BLD, ENH: Reading of extra flags from site.cfg to extend flexibility

zerothi opened the pull request:
BLD: Because it allows greater flexibility of the build phase for numpy, but no changes to _how_ numpy is build has been made

Addition to the distutils module to be able
to read in more optional arguments and flags
from the site.cfg file.

Up till this commit it has really been a _pain in the neck_ to optimize the build for certain flags, etc.
With this any user should be capable of providing their own flags/linker options to control the build
more easily.  
This commit can also be merged in the 1.9 maintenance branch.

Especially the runtime_library_dirs can come handy if users rely on LD_LIBRARY_PATH to super seed such env's.

There is still the caveat that hard-coded flags are not overwritten in any way so that providing additional flags has to be compatible with the hard-coded flags for the specific compiler.

Currently these additional options read:
- runtime_library_dirs:
  It allows users to set the runtime
  library directories so that LD_LIBRARY_PATH can be ignored.
  This has the same format as the library_dirs option.
- extra_compile_args:
  This allows the user to add specific compiler
  flags when compiling sources.
  There will be no formatting/checking of these additional compile
  flags, the compiler should complain if something is wrong.
- extra_link_args:
  This allows the user to add specific linker flags
  when linking the .so objects.
  There will be no formatting/checking of these additional compile
  flags, the linker should complain if something is wrong.

When the config is runned it automatically prints out the
read in information, thereby allowing the user to see
what has been set and what has not.

Tested with and without the added options to check that it builds correctly.

zerothi committed:
BLD, ENH: Reading of extra flags from site.cfg to extend flexibility

Addition to the distutils module to be able
to read in more optional arguments and flags
from the site.cfg file.

Currently are these additional options read:

runtime_library_dirs:
  It allows users to set the runtime
  library directories so that LD_LIBRARY_PATH can be ignored.
  This has the same format as the library_dirs option.

extra_compile_args:
  This allows the user to add specific compiler
  flags when compiling sources.
  There will be no formatting/checking of these additional compile
  flags, the compiler should complain if something is wrong.

extra_link_args:
  This allows the user to add specific linker flags
  when linking the .so objects.
  There will be no formatting/checking of these additional compile
  flags, the linker should complain if something is wrong.

When the config is runned it automatically prints out the
read in information, thereby allowing the user to see
what has been set and what has not.

Tested with and without flags to check that it builds correctly.

rgommers commented:
This looks interesting. As an enhancement, should go into master only though, not 1.9.x

Can you please add:
- tests, in ``distutils/tests/test_npy_pkg_config.py``
- some comments in ``site.cfg.example`` to explain how to use this functionality
- a mention in the 1.10 release notes (``doc/release/1.10.0-notes.rst``)

rgommers commented:
You may also want to ping the mailing list about this PR; probably there are a few people interested in testing this.

zerothi commented:
Thanks, will do during today. :)

2015-02-24 5:11 GMT+00:00 Ralf Gommers <notifications@github.com>:

> You may also want to ping the mailing list about this PR; probably there
> are a few people interested in testing this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597#issuecomment-75700573>.
>



-- 
Kind regards Nick

rgommers commented:
great. please ping once that's done - Github doesn't send notifications when only adding commits

zerothi commented:
ping @rgommers , requests has been added

However, the test is difficult to complete due to different architectures etc. For now it simply checks that the flags are read in by the system_info object in distutils.
I have added in the message log how to improve on this test, but I have not tried linking in distutils and thus I think it would not be wise for me to do.

zerothi committed:
ENH: Added tests to the extra options read in

A simple test (distutils/testing/test_system_info.py)
to check that the options are read in correctly has been added.
This test has a few faults:

A) It does not allow strict library checks as that can be
   _very_ system dependent.
B) It compiles some simple C-programs but does currently not link
   them to a shared library.
C) As such the test does not check that the flags are actually used.

To circumvent this one should:

A) Make a library of the compiled sources.
B) Check that a runtime_library_dirs is working by checking
   with ldd
C) Make a preprocessor flag to check the output of two commands which
   should differ according to the flags in each block

I am not too much into the distutils compiler suite. So I have not
endeavoured on this path.

- The current test shows that the flags are read in by the standard system_info
  object and can thus be considered a "stable" solution.

- Added note of the 1.10 release schedule.

- Corrected the site.cfg.example, added runtime_library_dirs to
  the OpenBLAS example where it seems appropriate.

- Bugfix for the site.cfg.example (the [DEFAULT] block should be
  name [ALL])
  This might have lead to some confusion, but many of the libraries
  are linked explicitly by their own sections, hence it might not have
  been caught.

zerothi committed:
BUG: Trying to fix travis builds

The test on Travis does not allow getcwd on python 2.6

The test for python 3+ already is in unicode, hence the
str representation was wrong.

zerothi committed:
BUG: Trying to fix travis builds

The error of getcwd on 2.6 was due to previous fault.
We should not request include_dirs as they are not provided.

For python 3 I forgot the top ascii writing.

rgommers commented:
@zerothi nice, those tests are quite a bit more extensive than what I expected! It'll take me some time to wrap my head around it, and test on a few other platforms than only 64-bit Linux (which is what TravisCI uses).

rgommers commented on doc/release/1.10.0-notes.rst:
```diff
@@ -74,6 +77,18 @@ Also, the dtype.type of nested structured fields is now inherited.
+  runtime_library_dirs (sets the runtime library directories to override
+    LD_LIBRARY_PATH)
+  extra_compile_args (add extra flags to the compilation of sources)
+  extra_link_args (add extra flags when linking libraries)
```
formatting: this should be a reST list, so ``-`` in front of each entry and blank lines above and below the list. and names with underscores should be surrounded by double backticks, otherwise Sphinx interprets those underscores as markup

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -489,8 +495,9 @@ def parse_config_files(self):
-        libs = self.get_libraries()
-        dirs = self.get_lib_dirs()
+        libs   = self.get_libraries()
+        dirs   = self.get_lib_dirs()
```
can you undo the added spaces here? PEP8 frowns upon such ``=`` alignment between lines

zerothi replied:
Ahh, sorry, I always do that in my own scripts (it makes me read lines
faster). I will correct those things.

I have never done a reST list. Will correct those as well.

Regarding the tests, feel free to ping me with questions if any arise. :)

2015-02-25 8:43 GMT+01:00 Ralf Gommers <notifications@github.com>:

> In numpy/distutils/system_info.py
> <https://github.com/numpy/numpy/pull/5597#discussion_r25322728>:
>
> > @@ -489,8 +495,9 @@ def parse_config_files(self):
> >                  self.cp.add_section(self.section)
> >
> >      def calc_libraries_info(self):
> > -        libs = self.get_libraries()
> > -        dirs = self.get_lib_dirs()
> > +        libs   = self.get_libraries()
> > +        dirs   = self.get_lib_dirs()
>
> can you undo the added spaces here? PEP8 frowns upon such = alignment
> between lines
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597/files#r25322728>.
>



-- 
Kind regards Nick

rgommers commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -0,0 +1,208 @@
+from tempfile import mkstemp, mkdtemp
+
+import distutils
+from numpy.testing import *
```
Can you import here only what's needed? Looks like ``TestCase`` and ``run_module_suite`` only now, but note that in general we prefer to use ``assert_`` functions from ``numpy.testing`` rather than ``nose`` functions.

zerothi replied:
I'll try and hunt down what is needed, I merely copy-pasted from the other
test in the same directory (and then added the system_info import).

Will correct, thanks for the heads up, and sorry for the trouble!

2015-02-25 8:47 GMT+01:00 Ralf Gommers <notifications@github.com>:

> In numpy/distutils/tests/test_system_info.py
> <https://github.com/numpy/numpy/pull/5597#discussion_r25322878>:
>
> > @@ -0,0 +1,208 @@
> > +from __future__ import division, absolute_import, print_function
> > +
> > +import os
> > +from tempfile import mkstemp, mkdtemp
> > +
> > +import distutils
> > +from numpy.testing import *
>
> Can you import here only what's needed? Looks like TestCase and
> run_module_suite only now, but note that in general we prefer to use
> assert_ functions from numpy.testing rather than nose functions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597/files#r25322878>.
>



-- 
Kind regards Nick

rgommers replied:
no need to apologize at all! 
everything I commented on so far is about minor details

rgommers commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -0,0 +1,208 @@
+            try:
+                self.search_static_first = self.cp.getboolean(self.section, 'search_static_first')
+            except: pass
+        assert isinstance(self.search_static_first, int)
```
Please use ``assert_`` from ``numpy.testing`` here. Plain ``assert``s get stripped when byte-compiling with ``-O``, so cannot be used in test code.

rgommers commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -0,0 +1,208 @@
+    Overrides the get_standard_file from system_info
+    """
+    tmpdir = mkdtemp()
+    filename = tmpdir + '/' + fname 
```
Use ``os.path.join``, this won't work on all platforms.

zerothi replied:
yes, just caught that :)

2015-02-25 9:05 GMT+01:00 Ralf Gommers <notifications@github.com>:

> In numpy/distutils/tests/test_system_info.py
> <https://github.com/numpy/numpy/pull/5597#discussion_r25323444>:
>
> > +    notfound_action:
> > +      0 - do nothing
> > +      1 - display warning message
> > +      2 - raise error
> > +    """
> > +    cl = {'temp1': test_temp1,
> > +          'temp2': test_temp2
> > +          }.get(name.lower(), test_system_info)
> > +    return cl()
> > +
> > +def get_standard_file(fname):
> > +    """
> > +    Overrides the get_standard_file from system_info
> > +    """
> > +    tmpdir = mkdtemp()
> > +    filename = tmpdir + '/' + fname
>
> Use os.path.join, this won't work on all platforms.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597/files#r25323444>.
>



-- 
Kind regards Nick

rgommers commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -0,0 +1,208 @@
+        defaults['extra_link_args'] = []
+        self.cp = ConfigParser(defaults)
+        self.files = []
+        self.files.extend(get_standard_file('site.cfg'))
```
This creates a tempdir with site.cfg in it that is not cleaned up after execution of the test. Should be done manually, like in ``TestSystemInfoReading.tearDown``.

rgommers commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -0,0 +1,208 @@
+
+def get_standard_file(fname):
+    """
+    Overrides the get_standard_file from system_info
```
maybe better to give this function a different name, overriding an imported name seems to not serve any purpose and precludes testing that imported function later.

zerothi replied:
Ok. I will try and work around it by incorporating it to the testcase
instead, seems easier to handle the teardown in the testcase anyway.

2015-02-25 9:15 GMT+01:00 Ralf Gommers <notifications@github.com>:

> In numpy/distutils/tests/test_system_info.py
> <https://github.com/numpy/numpy/pull/5597#discussion_r25323804>:
>
> > +
> > +def get_class(name, notfound_action=1):
> > +    """
> > +    notfound_action:
> > +      0 - do nothing
> > +      1 - display warning message
> > +      2 - raise error
> > +    """
> > +    cl = {'temp1': test_temp1,
> > +          'temp2': test_temp2
> > +          }.get(name.lower(), test_system_info)
> > +    return cl()
> > +
> > +def get_standard_file(fname):
> > +    """
> > +    Overrides the get_standard_file from system_info
>
> maybe better to give this function a different name, overriding an
> imported name seems to not serve any purpose and precludes testing that
> imported function later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597/files#r25323804>.
>



-- 
Kind regards Nick

rgommers commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -0,0 +1,190 @@
+        os.chdir(previousDir)
+
+    def test_compile2(self):
+        """ Compile source and link the second source """
```
Can you change all test docstrings to comments? Docstrings mess up the verbose output of ``nose``, i.e. it shows the docstring instead of ``test_file.test_name``. The latter is much easier to locate when searching where a test failure comes from.

rgommers commented on numpy/distutils/tests/test_system_info.py:
```diff
@@ -0,0 +1,190 @@
+
+class test_temp1(test_system_info):
+    section = 'temp1'
+class test_temp2(test_system_info):
```
PEP8 nitpick: two blank lines between each top-level def/class object

rgommers replied:
and another one: class names should be CamelCase

zerothi replied:
Ok, but what about test_system_info. It makes more sense/easier to follow
if they follow the naming in system_info? Or should I just change
everything to CamelCase?

2015-02-25 9:26 GMT+01:00 Ralf Gommers <notifications@github.com>:

> In numpy/distutils/tests/test_system_info.py
> <https://github.com/numpy/numpy/pull/5597#discussion_r25324231>:
>
> > +        self.parse_config_files()
> > +        if self.section is not None:
> > +            # Have to by-pass a boolean conversion for non-existing variable
> > +            try:
> > +                self.search_static_first = self.cp.getboolean(self.section, 'search_static_first')
> > +            except: pass
> > +        assert_(isinstance(self.search_static_first, int))
> > +
> > +    def _check_libs(self, lib_dirs, libs, opt_libs, exts):
> > +        """Override _check_libs to return with all dirs """
> > +        info = {'libraries' : libs , 'library_dirs' : lib_dirs }
> > +        return info
> > +
> > +class test_temp1(test_system_info):
> > +    section = 'temp1'
> > +class test_temp2(test_system_info):
>
> and another one: class names should be CamelCase
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597/files#r25324231>.
>



-- 
Kind regards Nick

rgommers replied:
I'd say rename to TestSystemInfo as well. ``distutils`` is a bit outdated, but this is a completely new file so can be done the "modern" way as a whole I'd say.

rgommers replied:
but you're right, inside the existing codebase it's better to conform to what's there now.

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -498,17 +505,46 @@ def calc_libraries_info(self):
+            opt = self.cp.get(self.section, key)
+            if opt:
+                tmp = { key : [opt] }
+                dict_append(info,**tmp)
```
a few more PEP8 nitpicks: space after comma on the line above and on the line ``for key in ...``. And no spaces around ``key : [opt]`` two lines above.

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -498,17 +505,46 @@ def calc_libraries_info(self):
+                # runtime_library_dirs
+                del i['libraries']
+                i['runtime_library_dirs'] = i.pop('library_dirs')
+                dict_append(info, **i)
```
The above looks correct, but it took me a couple of minutes to understand the comment and the use of ``del``. How about replacing the whole if clause with:

    # From `check_libs` we get keys "libraries" and "library_dirs",
    # while we need only key "runtime_library_dirs" with those paths
    dict_append(info, runtime_library_dirs=i['library_dirs'])

zerothi replied:
It was more to be consistent in that if later revisions of check_libs
returns more than the two keys (libraries and library_dirs). In that
_future_ case would any additional keys also get passed. Hence the obscure
delete, pop sequence.
It may be overkill?

2015-02-25 9:53 GMT+01:00 Ralf Gommers <notifications@github.com>:

> In numpy/distutils/system_info.py
> <https://github.com/numpy/numpy/pull/5597#discussion_r25325312>:
>
> > @@ -498,17 +505,46 @@ def calc_libraries_info(self):
> >                  dict_append(info, **i)
> >              else:
> >                  log.info('Library %s was not found. Ignoring' % (lib))
> > +            i = self.check_libs(r_dirs, [lib])
> > +            if i is not None:
> > +                # Swap library keywords found to runtime_library_dirs
> > +                # the libraries are insisting on the user having defined
> > +                # them using the library_dirs, and not necessarily by
> > +                # runtime_library_dirs
> > +                del i['libraries']
> > +                i['runtime_library_dirs'] = i.pop('library_dirs')
> > +                dict_append(info, **i)
>
> The above looks correct, but it took me a couple of minutes to understand
> the comment and the use of del. How about replacing the whole if clause
> with:
>
> # From `check_libs` we get keys "libraries" and "library_dirs",
> # while we need only key "runtime_library_dirs" with those paths
> dict_append(info, runtime_library_dirs=i['library_dirs'])
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597/files#r25325312>.
>



-- 
Kind regards Nick

rgommers replied:
Then all those extra keys would already be updated by this code above right?

            i = self.check_libs(dirs, [lib])
            if i is not None:
                dict_append(info, **i)

zerothi replied:
Yes, but that would also add the libraries key, which should be provided by the library key and library_dirs, not intrinsically by runtime_library_dirs. I did this so that library_dirs has precedence for actual libraries, and runtime_library_dirs only works for those directories listed.
So it is a "hack" to not re-create a check_libs routine only for the runtime libs.
The way I did it it will only allow runtime_library_dirs to be added iff the libraries also exists in those paths (this is what check_libs does).

If your above code would be used, all libraries would be added twice to the list. Also we need to transfer library_dirs to runtime_library_dirs, hence the pop to the other key.

###### site1.cfg
```
[lapack]
libraries = lapack
library_dirs = /opt/lapack
```
Linking command: `... -L/opt/lapack -llapack`.  
Here `LD_LIBRARY_PATH` is needed if shared lapack is shared.
###### site2.cfg
```
[lapack]
libraries = lapack
library_dirs = /opt/lapack
runtime_library_dirs = /opt/lapack
```
Linking command: `... -L/opt/lapack -Wl,-rpath=/opt/lapack -llapack`.  
Here the runtime library directory will be set, `LD_LIBRARY_PATH` not needed any more.
###### site3.cfg
```
[lapack]
libraries = lapack
runtime_library_dirs = /opt/lapack
```
Linking command: `... -Wl,-rpath=/opt/lapack -llapack`.

rgommers commented on site.cfg.example:
```diff
@@ -52,6 +52,26 @@
+#
+#   runtime_library_dirs
+#       List of directories that contains the libraries that should be 
+#       used at runtime, thereby disregarding the LD_LIBRARY_PATH variable.
```
it's more "removing the need to use" rather then "disregarding" right? One can still use a combination of both ``runtime_library_dirs`` and ``LD_LIBRARY_PATH`` if desired.

zerothi replied:
It depends, you cannot override a set runtime_library_dir by
LD_LIBRARY_PATH. Only if the runtime library directory has not been set.
Hence, disregarding.

2015-02-25 10:10 GMT+01:00 Ralf Gommers <notifications@github.com>:

> In site.cfg.example
> <https://github.com/numpy/numpy/pull/5597#discussion_r25326065>:
>
> > @@ -52,6 +52,26 @@
> >  #       True) to tell numpy.distutils to prefer static libraries (.a) over
> >  #       shared libraries (.so). It is turned off by default.
> >  #           search_static_first = false
> > +#
> > +#   runtime_library_dirs
> > +#       List of directories that contains the libraries that should be
> > +#       used at runtime, thereby disregarding the LD_LIBRARY_PATH variable.
>
> it's more "removing the need to use" rather then "disregarding" right? One
> can still use a combination of both runtime_library_dirs and
> LD_LIBRARY_PATH if desired.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597/files#r25326065>.
>



-- 
Kind regards Nick

rgommers replied:
true. So yes, maybe better leave as is.

zerothi committed:
BUG: PEP corrections and streamlined with numpy dev line

Changed self.assert* to assert_ instances through numpys
own testing utilities.

Fixes for the rst document.

Removed unnecessary import statements in the test.

rgommers commented:
For the extra compile flags, it might be useful to add a comment to point out that this does what now goes into ``'ignored'`` for a ``[cflags]`` section in site.cfg.

rgommers commented:
I still have to go through this with a debugger (the logic in distutils doesn't fit in my head) but can you confirm that ``runtime_library_dirs`` does fill the existing variables of the same name in ``ccompiler.py``, ``build_ext.py``, ``extension.py`` and ``fcompiler/__init__.py``?

rgommers commented:
Overall I think the logic in this patch looks good, and ``system_info`` changes (which are always fragile) are kept as minimal as possible.

zerothi commented:
I think you are thinking of the pkg reading of settings. Those are two
separate entities. The pkg information and site.cfg does not interact in
any way (to my knowledge?).

2015-02-25 10:14 GMT+01:00 Ralf Gommers <notifications@github.com>:

> For the extra compile flags, it might be useful to add a comment to point
> out that this does what now goes into 'ignored' for a [cflags] section in
> site.cfg.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597#issuecomment-75928191>.
>



-- 
Kind regards Nick

zerothi committed:
BUG: PEP corrections

More corrections pointed out by Ralf

Changed the get_standard_file to a fully temporary
file. This means that the __init__ diverges a bit
from the system_info object. However, it only has
to do with the setup for the test. All internal things
regarding the object have not been altered.
I have checked on my box that all files/directories are removed.

zerothi committed:
BUG: Python3 fix

Direct writing to files is not the same for 2vs3.
Hence we now close the file handle, re-open it
and write using the file.

zerothi commented:
Yes and no, it fully depends on how the extensions/compilations are
created. What I have done only adds the flags to separate sections (and the
default one).

Example (lapack_lite):
>
lapack_info = get_info('lapack_opt', 0)
config.add_extension('lapack_lite',
                         sources = [get_lapack_lite_sources],
                         depends = ['lapack_litemodule.c'] +
lapack_lite_src,
                         extra_info = lapack_info
                         )
>
in this case the compilation of the sources [get_lapack_lite_sources] will
be done with the extra stuff in the lapack section (and hence will be
retrieved in the ccompiler etc. sources).

If I wanted a generic extension to take advantage of the ALL section in
site.cfg then I needed this:
>
all_info = get_info('i_just_want_the_default', 0) # a non-existing class
name returns the ALL section
config.add_extension('my_package',
                         sources = ['my_package.c'],
                         extra_info = all_info
                         )
>
Reading of the flags are not intrinsic to the compiler, nor the
configuration class, they need to be passed in some way. :(

If you want to compile a file with some flags from site.cfg you need
something like the test I created:
>
all_info = {'extra_link_args':[],'extra_compile_args':[]} # ensures default
of nothing
all_info.update(get_info('i_just_want_the_default', 0))
c.compile(['source.c'], output_dir='/home/user',

extra_postargs=all_info['extra_link_args']+all_info['extra_compile_args'])
>
Hence the answer is that it depends on how the user creates the
compilation, in any case you still need to create the info-object to
retrieve the flags.
I hope this answers your question?

2015-02-25 10:16 GMT+01:00 Ralf Gommers <notifications@github.com>:

> I still have to go through this with a debugger (the logic in distutils
> doesn't fit in my head) but can you confirm that runtime_library_dirs
> does fill the existing variables of the same name in ccompiler.py,
> build_ext.py, extension.py and fcompiler/__init__.py?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597#issuecomment-75928433>.
>



-- 
Kind regards Nick

zerothi commented:
@rgommers I have corrected a few things and added `rpath` to be equivalent to `runtime_library_dirs`. As I dug deeper in the pure distutils package I saw that `rpath` is the way to add `runtime_library_dirs`, yet the keywords internally are `runtime_library_dirs`. This simple abstraction I think is ok as it is easily circumvented. 
Note, the test also tests the `rpath` key.

Please advice if anything needs added or tested.

zerothi committed:
ENH: rpath and runtime_library_dirs equivalent

The original distutils assumes runtime_library_dirs to
be located in rpath, however, the internal structures assumes
the keyword to be runtime_library_dirs.
For now numpy.distutils handles both equivalently.

The test has been updated to also test the rpath solution.

zerothi committed:
BUG: Had thought the defaults where made of lists, it is simply a str.

Changed all lists to strings

zerothi committed:
BUG: rpath was not implemented for numpy.distutils.fcompiler

This bug-fix only applies for non gnu-compilers.

The fortran compilers in numpy inherited the Ccompier class
which had the runtime library directories NotImplemented.
Hence the compilers need to define their own runtime library
path.
I have added that information to the intel/pgi/sun compilers.
The rest are either already implemented or I do not know them.

I have tested the bug-fix with intel compilers of numpy AND the
subsequent installation of scipy (which relies on the fortran
compilers). Hence this bug will only occur when linking with
the fortran compilers.

juliantaylor commented on site.cfg.example:
```diff
@@ -52,6 +52,28 @@
+#       Add additional arguments to the compilation of sources.
+#       Simple variable with no parsing done. 
+#       Provide a single line with all complete flags.
+#           extra_compile_args = -g -ftree-vectorize
```
it is not clear here what the scope of these arguments are, library/runtime/include-dir are all related to binaries, compile args apply to sources. Does it apply to all numpy sources like the OPT env variable) or only part of them?
I don't see a reason this flag should be specific to sections concerning libraries, maybe having one global one is enough?

zerothi replied:
My addition _only_ applies to codes/extensions using the `get_info` function from `numpy.distutils`. So I can't say if it is used for all numpy sources like `OPT`. See my comment on `lapack_lite` which shows one place where it would be used. So, these compile flags are _not_ used unless `get_info` is called on a section containing the above mentioned args.

I have added the `extra_compile_args` mainly for completeness sake, I see no disadvantage of having the possibility of adding extra flags when compiling extensions. You may say it is not necessary for libraries, but it allows people to construct sections which can be used to compiling extensions with different flags, say:

```
[O2]
extra_compile_args = -O2
[O3]
extra_compile_args = -O3
```

And then compiling can be done using:
```
config.add_extension('my_package',
                         sources = ['my_package.c'],
                         extra_info = get_info('O2')
                         )
config.add_extension('my_package3',
                         sources = ['my_package3.c'],
                         extra_info = get_info('O3')
                         )
```
One can of course extend several sections together.  
I like to be able to fully control my compilation of source when using Python, with this PR the user can add new sections to their `site.cfg` and then control all sources/extensions individually, if needed. 

Furthermore using config files, I think, is much more streamlining the build process than having to use env's. This is a matter of taste, but I see no disadvantage in letting users have a choice of this.

zerothi replied:
But mainly I would use it for the `rpath` options to control the shared libraries. :)

juliantaylor commented:
thanks this looks good, especially nice that you went the extra mile and added the tests for this not very well tested area. I'll try to do some testing with the branch soon.

charris commented:
@jaimefrio @rgommers What is the status of this?

zerothi commented:
I am still following this in case you need anything done.
It seems you are pushing for a soon 1.10 release?

2015-03-22 23:24 GMT+01:00 Charles Harris <notifications@github.com>:

> @jaimefrio <https://github.com/jaimefrio> @rgommers
> <https://github.com/rgommers> What is the status of this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597#issuecomment-84716667>.
>



-- 
Kind regards Nick

charris commented:
Starting to prepare, yes. I was waiting for @juliantaylor, but this seems to be ready.

zerothi commented:
Ok, great. :)

2015-04-07 19:14 GMT+02:00 Charles Harris <notifications@github.com>:

> Starting to prepare, yes. I was waiting for @juliantaylor
> <https://github.com/juliantaylor>, but this seems to be ready.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5597#issuecomment-90650800>.
>



-- 
Kind regards Nick

charris commented:
@juliantaylor I'm about ready to put this in. Do you want to do any testing?

charris commented:
OK, putting this in. @juliantaylor can more easily test it now ;)
